### PR TITLE
BUG: FIX empty slice error + add test

### PR DESCRIPTION
**Pull request by kingjr**
A bug was introduced in the previous optimization [PR](https://github.com/mne-tools/mne-python/pull/2673) if the cross val object had some empty `test` it generates an empty slide an do not have anything to predict anymore.

If possible, backpropagate (that's not the right term...) this to the stable version?

### CODE REVIEW

File: mne/decoding/tests/test_time_gen.py [hunk 1]
```diff
@@ -241,6 +241,10 @@ def test_generalization_across_time():
 
     # TODO JRK: test GAT with non-exhaustive CV (eg. train on 80%, test on 10%)
 
+    # Make CV with some empty tests to ensure that the folds are passed
+    gat.cv_.test_folds[gat.cv_.test_folds == 1] = 0
+    gat.predict(epochs)
```
Comment (dengemann) [inline mne/decoding/tests/test_time_gen.py]: when would you use this scenario? For bootstrapping?

Comment (kingjr) [inline mne/decoding/tests/test_time_gen.py]: I encountered this because I misused a stratified k fold with continuous `y`, and it generated empty test folds. It's obviously not a recommended scenario, but the `gat` should not crash because of an empty test fold any way.

Shall I also add a warning to check if there's an empty test fold in the cv?

Comment (Eric89GXL) [inline mne/decoding/tests/test_time_gen.py]: I don't know much about this, but I agree it's important to have empty cases handled gracefully -- it often comes up during parallelization if you have more jobs/cores than slices of data to operate on. This sounds slightly different, but the idea I think is the same, that it shouldn't bomb. It does sound to me like having empty folds is probably not supposed to happen in this case, though -- if that's true and it's likely indicative of some user error, +1 for warning (or error if it should really never happen) from me.

File: mne/decoding/time_gen.py [hunk 2]
```diff
@@ -449,6 +449,8 @@ def _predict_time_loop(X, estimators, cv, slices, predict_mode):
         elif predict_mode == 'cross-validation':
             # Predict with the estimator trained on the separate training set.
             for k, (train, test) in enumerate(cv):
+                if not np.any(test):
```
Comment (Eric89GXL) [inline mne/decoding/time_gen.py]: Although maybe the check here is too lenient. In other words, `test` could be non-empty and all zeros and this will evaluate to go to the `continue`, which isn't right I don't think. Maybe you want to do `if np.asarray(test).size == 0`?

File: mne/decoding/tests/test_time_gen.py (hunk 1 above)
Comment (kingjr) [inline mne/decoding/tests/test_time_gen.py]: Ok I'll add a warning for empty test fold (which isn't recommended but shouldn't create an error) and raise an error for empty train fold (in which case we can't fit the estimator)

**Comment by kingjr** [comment]
Ready on my part.

File: mne/decoding/tests/test_time_gen.py [hunk 3]
```diff
@@ -241,6 +241,15 @@ def test_generalization_across_time():
 
     # TODO JRK: test GAT with non-exhaustive CV (eg. train on 80%, test on 10%)
 
+    # Make CV with some empty train and test folds:
+    # --- empty test fold(s) should warn when gat.predict()
+    gat.cv_.test_folds[gat.cv_.test_folds == 1] = 0
+    with warnings.catch_warnings(record=True):
+        gat.predict(epochs)
```
Comment (agramfort) [inline mne/decoding/tests/test_time_gen.py]: you're not actually testing that a warning was raised.

File: mne/decoding/time_gen.py [hunk 4]
```diff
@@ -449,6 +454,8 @@ def _predict_time_loop(X, estimators, cv, slices, predict_mode):
         elif predict_mode == 'cross-validation':
             # Predict with the estimator trained on the separate training set.
             for k, (train, test) in enumerate(cv):
+                if np.asarray(test).size == 0:
```
Comment (agramfort) [inline mne/decoding/time_gen.py]: test should already be an array. Just do `test.size == 0`

**Comment by kingjr** [comment]
I don't manage to catch the warning. It's originally showing up, but it then disappears despite `warnings.simplefilter('always')`. Help welcome...

**Comment by agramfort** [comment]
you need to add at the top of the test file:

`warnings.simplefilter('always')  # enable b/c these tests throw warnings``

then you can do:

```
    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter('always')
        gat.predict(epochs)
    assert_true(len(w) == 1)
    assert_true('Some folds do not have any test epochs.' in w[0].message)
```

please see if you can get rid of all the warnings in tests

**Comment by kingjr** [comment]
> please see if you can get rid of all the warnings in tests

By catching them?

**Comment by agramfort** [comment]
yes

**Comment by kingjr** [comment]
Is it possible to catch the warning sent through `logger.warning`? Sorry to be a pain...

**Comment by agramfort** [comment]
no idea

see if it's done elsewhere in the code.

**Comment by kingjr** [comment]
> see if it's done elsewhere in the code.

Didn't find. Could have searched better.

**Comment by kingjr** [comment]
Ok, good for reviews

**Comment by kingjr** [comment]
Hopefully I managed the rebase...

File: mne/decoding/tests/test_time_gen.py [hunk 5]
```diff
@@ -205,33 +206,45 @@ def test_generalization_across_time():
 
     # Test testing time parameters
     # --- outside time range
-    gat.test_times = dict(start=-999.)
-    assert_raises(ValueError, gat.predict, epochs)
-    gat.test_times = dict(start=999.)
-    assert_raises(ValueError, gat.predict, epochs)
+    with warnings.catch_warnings(record=True) as w:
+        # This scenario use has len(y) < n_folds and thus throws a warning
```
Comment (jasmainak) [inline mne/decoding/tests/test_time_gen.py]: you need to put two `warnings.simplefilter`. One at the top of the file and one in each `warnings.catch_warnings`. Look at how we have done it elsewhere

**Comment by agramfort** [comment]
merged by rebase

thx @kingjr
